File: apps/backend/src/tools/social-validation-engine/analyzers/keyword-generator.ts
/**
 * analyzers/keyword-generator.ts
 *
 * Generates 5–10 problem-framed search keywords from a startup idea using Gemini.
 * These keywords drive the Reddit / HN / Product Hunt collectors.
 */

import { requireGeminiClient, withTimeout, parseRetryDelay } from '../../../services/gemini';
import { SETTINGS } from '../config';

const SYSTEM_PROMPT = `Generate 5 short, problem-focused search query phrases for this startup idea. Output JSON: {"keywords": ["string"]}`;

export async function generateKeywords(ideaText: string): Promise<string[]> {
  const ai = requireGeminiClient();

  let lastExc: any = null;

  for (let attempt = 0; attempt < 3; attempt++) {
    try {
      const response = await withTimeout(
        ai.models.generateContent({
          model: SETTINGS.geminiModel,
          contents: `Idea: ${ideaText}`,
          config: {
            systemInstruction: SYSTEM_PROMPT,
            maxOutputTokens: 128,
            responseMimeType: 'application/json',
            responseSchema: {
              type: 'OBJECT',
              properties: {
                keywords: {
                  type: 'ARRAY',
                  items: { type: 'STRING' }
                }
              },
              required: ['keywords']
            }
          }
        }),
        SETTINGS.keywordGenTimeoutMs,
        'keyword_generator'
      );

      const raw = response.text?.trim() || '';
      const data = JSON.parse(raw);
      const keywords: string[] = data.keywords || [];
      if (keywords.length === 0) throw new Error('Gemini returned an empty keyword list.');
      console.log(`keyword_generator: generated ${keywords.length} keywords.`);
      return keywords.slice(0, 10);

    } catch (exc: any) {
      lastExc = exc;
      const msg = String(exc.message || exc);
      const delay = parseRetryDelay(msg);
      console.warn(`keyword_generator: attempt ${attempt + 1} failed — ${msg.substring(0, 120)}. Retrying in ${delay / 1000}s…`);
      await new Promise(r => setTimeout(r, delay));
    }
  }

  throw new Error(`keyword_generator failed after 3 attempts. Last: ${lastExc?.message || lastExc}`);
}

File: apps/backend/app/tools/sve/analyzers/keyword_generator.py
import json
import asyncio
from google.genai import types
from app.services.gemini_client import require_gemini_client
from app.tools.sve.config import SETTINGS

SYSTEM_PROMPT = "Generate 5 short, problem-focused search query phrases for this startup idea. Output JSON: {\"keywords\": [\"string\"]}"

async def generate_keywords(idea_text: str) -> list[str]:
    ai = require_gemini_client()
    last_exc = None

    for attempt in range(3):
        try:
            # Call Gemini using async client.aio namespace
            response = await ai.aio.models.generate_content(
                model=SETTINGS["geminiModel"],
                contents=f"Idea: {idea_text}",
                config=types.GenerateContentConfig(
                    system_instruction=SYSTEM_PROMPT,
                    max_output_tokens=128,
                    response_mime_type="application/json",
                    response_schema={
                        "type": "OBJECT",
                        "properties": {
                            "keywords": {
                                "type": "ARRAY",
                                "items": {"type": "STRING"}
                            }
                        },
                        "required": ["keywords"]
                    }
                )
            )

            raw = (response.text or "").strip()
            data = json.loads(raw)
            keywords = data.get("keywords") or []
            if not keywords:
                raise ValueError("Gemini returned an empty keyword list.")
            print(f"keyword_generator: generated {len(keywords)} keywords.")
            return keywords[:10]
        except Exception as e:
            last_exc = e
            print(f"keyword_generator: attempt {attempt + 1} failed: {e}. Retrying in 2s...")
            await asyncio.sleep(2.0)

    raise RuntimeError(f"keyword_generator failed after 3 attempts. Last error: {last_exc}")
